sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,315 @@
+            is_inlier = np.ones(X.shape[0], dtype=int)
+            is_inlier[self.decision_function(X) <= self.threshold_] = -1
+        else:
+            is_inlier = np.ones(self._fit_X.shape[0], dtype=int)
+            is_inlier[-self.outlier_factor_ <= self.threshold_] = -1
+
+        return is_inlier
+
+    def decision_function(self, X):
+        """Opposite of the Local Outlier Factor of X (as bigger is better).
+
+        The argument X is supposed to contain *new data*: if X contains a
+        point from training, it consider the later in its own neighborhood.
+        Also, the samples in X are not considered in the neighborhood of any
+        point.
+        The decision function on training data is available by considering the
+        opposite of the outlier_factor_ attribute.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples.
+
+        Returns
+        -------
+        lof_scores : array, shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_distances_fit_X_",
+                               "_neighbors_indices_fit_X_"])
+
+        X = check_array(X, accept_sparse='csr')
+
+        distances_X, neighbors_indices_X = self.kneighbors(X)
+        # as bigger is better:
+        return -self._local_outlier_factor(distances_X, neighbors_indices_X)
+
+    def _local_outlier_factor(self, distances_X, neighbors_indices_X):
+        """Compute the local outlier factor (LOF)
+
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        distances_X : array, shape (n_query, self.n_neighbors)
+            Distances to the neighbors (in the training samples self._fit_X) of
+            each query point to compute the LRD.
+
+        neighbors_indices : array, shape (n_query, self.n_neighbors)
+            Neighbors indices (of each query point) among training samples
+            self._fit_X.
+
+        Returns
+        -------
+        lof : numpy array, shape (n_samples,)
+            The LOF of X. The lower, the more normal.
+        """
+        n_samples = distances_X.shape[0]
+
+        # Compute the local_reachability_density of samples X:
+        X_lrd = self._local_reachability_density(distances_X,
+                                                 neighbors_indices_X)
+        lrd_ratios_array = np.zeros((n_samples, self.n_neighbors))
+
+        # Avoid re-computing X_lrd if same parameters:
+        if not (np.all(distances_X == self._distances_fit_X_) *
+                np.all(self._neighbors_indices_fit_X_ == neighbors_indices_X)):
```

Comment:
this ``==`` raises a deprecation warning

> /home/andy/checkout/scikit-learn/sklearn/neighbors/lof.py:279: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
  np.all(self._neighbors_indices_fit_X_ == neighbors_indices_X)):

This means they have different size, I think. So I guess you should check the shape first?

Also happens for the line above.